pkg/content/provider/git/git.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/GoogleContainerTools/kpt/internal/util/git"
+	"github.com/GoogleContainerTools/kpt/internal/util/remote"
+	"github.com/GoogleContainerTools/kpt/pkg/content"
+	"github.com/GoogleContainerTools/kpt/pkg/content/extensions"
+	"github.com/GoogleContainerTools/kpt/pkg/content/provider/dir"
+	"github.com/GoogleContainerTools/kpt/pkg/location"
+	"github.com/GoogleContainerTools/kpt/pkg/location/mutate"
+)
+
+type base interface {
+	extensions.FileSystemProvider
+	extensions.FSProvider
+}
+
+type gitProvider struct {
+	base
+
+	repoSpec *git.RepoSpec
+}
+
+var _ content.Content = &gitProvider{}
+var _ extensions.FileSystemProvider = &gitProvider{}
+var _ extensions.FSProvider = &gitProvider{}
+
+func Open(ctx context.Context, ref location.Git) (content.Content, location.ReferenceLock, error) {
+	repoSpec := &git.RepoSpec{
+		OrgRepo: ref.Repo,
+		Path:    filepath.Join("/", ref.Directory),
+		Ref:     ref.Ref,
+	}
+
+	return open(ctx, ref, repoSpec)
+}
+
+func OpenLock(ctx context.Context, ref location.GitLock) (content.Content, location.ReferenceLock, error) {
+	repoSpec := &git.RepoSpec{
+		OrgRepo: ref.Repo,
+		Path:    filepath.Join("/", ref.Directory),
+		Ref:     ref.Commit,
+	}
+
+	return open(ctx, ref, repoSpec)
+}
+
+func open(ctx context.Context, ref location.Reference, repoSpec *git.RepoSpec) (_result content.Content, _ location.ReferenceLock, _err error) {
+
+	if err := remote.ClonerUsingGitExec(ctx, repoSpec); err != nil {
```

Comment:
Yep, exactly like you mentioned earlier, in the `c, err := open.Content(ref, ...options)` an option will need to enable getting in front of the built in switch cases - and for new public api that transitively and recursively opens upstream packages I assume it would also take the opts ...open.Option at the calling point? e.g. to carry WithContext(ctx), any credentials for built-in openers, any custom openers added by the caller, etc.